pact/Pact/Core/IR/Eval/CEK.hs

```diff
@@ -731,30 +730,41 @@ acquireModuleAdmin i cont handler env mdl = do
    -> Eval (EvalResult CEKBigStep CoreBuiltin () Eval)
     #-}
 
+-- | Evaluate a term with all the stack manipulation logic
 evalWithStackFrame
   :: (CEKEval step b i m, MonadEval b i m)
   => i
   -> Cont step b i m
   -> CEKErrorHandler step b i m
   -> CEKEnv step b i m
-  -> StackFrame
   -> Maybe Type
+  -> StackFrame i
   -> EvalTerm b i
   -> m (CEKEvalResult step b i m)
-evalWithStackFrame info cont handler env sf mty body = do
+evalWithStackFrame info cont handler env mty sf body = do
   cont' <- pushStackFrame info cont mty sf
   evalCEK cont' handler env body
 
+-- | Push a stack frame into the stack, and check it for recursion
 pushStackFrame
   :: (MonadEval b i m)
   => i
   -> Cont step b i m
   -> Maybe Type
-  -> StackFrame
+  -> StackFrame i
   -> m (Cont step b i m)
 pushStackFrame info cont mty sf = do
+  checkRecursion
   esStack %== (sf :)
   pure (StackPopC info mty cont)
+  where
+  checkRecursion = do
+    RecursionCheck currentCalled <- usesEvalState esCheckRecursion NE.head
+    let qn = fqnToQualName (_sfName sf)
+    when (S.member qn currentCalled) $ throwExecutionError info (RuntimeRecursionDetected qn)
```

Comment:
So correct me if i'm wrong, but the process is as follows: 

When we see a fqn that's new, we check the current frame calling context to see if there's already a reference to it anywhere, treating the `NonEmpty RecursionCheck` state as a second stack that tracks names that it's already seen, and if it has been seen already, abort, but if not, load it into the associated recursion tracking context if it's not. 

This could get pretty costly. We'll need to find a way to gas this because any significantly complicated `defun` is going to have a lot of recursion checks no?

Reply:
My gut feel is this shouldn't be too costly relative to the other stuff CEK machine is doing, but it'd make a great benchmark case! I can take a look at that separately soon-ish.

Reply:
> This could get pretty costly.

Why would this be costly? The set will not grow very large in practice, and it just checks a qualified name. 

> We'll need to find a way to gas this because any significantly complicated defun is going to have a lot of recursion checks no?

We already charge an overhead for function application. The overhead covers the cost of a few things anyway:
- Pushing the arguments onto the environment
- Pushing a stackframe into the context stack
- Recursion checking.

In practice, that set will be small, ankd even if it were 10000 functions deep, it would not be slow enough for it to matter in the grand scheme of things, I believe, but we will simply use the benchmarks for function app to verify this 👍